include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child classes derived from the
+ * base DataType class to store data on a given cell. This implies the usage of pointers,
+ * in our case -- std::shared_ptr().
+ *
+ * @note Note that the data type stored on each cell can be different.
+ * However, within the cell this class stores a vector of objects of a single data type.
+ * For this reason, this class may not be sufficiently flexible when, for example,
+ * adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CellIteratorType, typename DataType>
+class CellDataStorage
+{
+public:
+  /**
+   * A typedef for a vector of shared pointers with arbitrary data type @p T .
+   */
+  template <typename T> using CellVector = std::vector<std::shared_ptr<T> >;
+
+  /**
+   * Default constructor.
+   */
+  CellDataStorage();
+
+  /**
+   * Default virtual destructor.
+   */
+  virtual ~CellDataStorage();
+
+  /**
+   * Initialize data on the @p cell to store @p number_of_data_points of objects of type @p T .
+   * The typename @p T is possibly another class which is derived from the
+   * base @p DataType class. In order to initialize the vector of objects
+   * we have to assume that the class @p T has a default constructor.
+   * This function has to be called on every cell where data is to be stored.
+   *
+   * After the data is initialized, it can be modified using the get_data()
+   * function.
+   *
+   * @note Subsequent calls of this function with the same @p cell will not
+   * alter the objects associated with it. In order to remove the stored data,
+   * use the erase() function.
+   *
+   * @note It is possible to use different types @p T for different cells which
+   * may reflect, for example, different constitutive models of continuum mechanics
+   * in different parts of the domain.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int number_of_data_points);
+
+  /**
+   * Same as above but for a range of iterators starting at @p cell_start
+   * until, but not including, @p cell_end for all locally owned cells, i.e.
+   * for which `is_locally_owned()==true`
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell_start,
+                  const CellIteratorType &cell_end,
+                  const unsigned int number);
+
+  /**
+   * Removes data stored at the @p cell. Returns true if the data was removed.
+   * If no data is attached to the @p cell, this function will not do anything
+   * and returns false.
+   *
+   * @note In debug mode, this function will also check that the are no
+   * outstanding references to the data stored on this cell. That is to say,
+   * that the only references to the stored data are that made by this class.
```

Comment:
This comment is odd because it exposes an internal detail, namely how the class stores its data as `shared_ptr`s. If a user had taken out a different shared ptr to this data, would it not be entirely reasonable that the other place is then responsible for (automatic) destruction? That's what `shared_ptr`s are all about after all. If you think that's not the appropriate semantics, maybe `unique_ptr` is for you :-)